### PR TITLE
add weight_sum_fp32 config

### PR DESCRIPTION
**Pull request by suexu1025**
# Description

Add config flag `weight_sum_fp32 ` for whether to use full fp32 precision for weight_sum during final unpermute in moe
# Tests

final eval loss at 300 steps
2.394 (cloudlog)(https://cloudlogging.app.goo.gl/Q5o2tac9aypGGMyV6)
2.393 (cloudlog)(https://cloudlogging.app.goo.gl/L2N43dAZiHap1Djk7)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/moe.py [hunk 1]
```diff
@@ -579,10 +579,13 @@ def unpermute(
...
         # For Llama4, combine using weights of 1 for selected experts
         reshaped_weights = jnp.ones_like(reshaped_weights)
+      if self.config.weight_sum_fp32:
+        reshaped_intermediate = reshaped_intermediate.astype(jnp.float32)
+        reshaped_weights = reshaped_weights.astype(jnp.float32)
       output = jnp.einsum(
           "BKE,BK -> BE",
-          reshaped_intermediate.astype(jnp.float32),
-          reshaped_weights.astype(jnp.float32),
+          reshaped_intermediate,
```
Comment (RissyRan) [inline src/MaxText/layers/moe.py:587]: Could you also help update this block at line #1616?

```
      with jax.named_scope("w_sum"):
        if is_llama4_decoder_layer:
          weights = self.reshape_and_update_weights(jnp.ones_like(top_k_weights), top_k_indices)
        # cast to f32 for sum up in einsum op
        output = jnp.einsum(
            "BSEM,BSE -> BSM",
            intermediate_layer.astype(jnp.float32),
            weights.astype(jnp.float32),  # pylint: disable=undefined-variable,possibly-used-before-assignment
            precision=matmul_precision,
        ).astype(self.dtype)
```

Comment (suexu1025) [reply src/MaxText/layers/moe.py:587]: Thanks, just added that part as well

File: src/MaxText/configs/base.yml [hunk 2]
```diff
@@ -149,6 +149,7 @@ logits_dot_in_fp32: False  # whether to use fp32 in logits_dense or shared_embed
 cast_logits_to_fp32: True # whether to cast the logits to fp32. The higher precision is generally beneficial, but it can vary slightly.
 float32_qk_product: False # in dot_product attention, whether to cast to fp32 the inputs to qk product
 float32_logits: False # in dot_product attention, whether to cast to fp32 the inputs to softmax
+weight_sum_fp32: True # whether to use full fp32 precision for weight_sum during final unpermute in moe
```
Comment (RissyRan) [inline src/MaxText/configs/base.yml]: shall we rename it to `float32_weight_sum` for consistency?

File: src/MaxText/layers/moe.py [hunk 3]
```diff
@@ -1673,11 +1676,14 @@ def dense_matmul(
       with jax.named_scope("w_sum"):
```
Comment (RissyRan) [inline src/MaxText/layers/moe.py]: Extra: That will be great if you could help update this to `weight_sum`

Comment (suexu1025) [reply src/MaxText/layers/moe.py]: updated

File: src/MaxText/configs/base.yml (hunk 2 above)
Comment (suexu1025) [reply src/MaxText/configs/base.yml]: udpated
